### PR TITLE
tls: ban SSL3, TLS1, TLS1.1 and DTLS1.0 at security level one and above

### LINKED ISSUE #18194: AES128-SHA tls1 still works with openssl3 seclevel=1
in NEWS.md you say
```
SSL 3, TLS 1.0, TLS 1.1, and DTLS 1.0 only work at security level 0.
```
this news message was added here:
https://github.com/openssl/openssl/commit/aba03ae571ea677fc484daef00a21ca8f7e82708
in result of degrading sha1

by accident I discovered that contrary to my expectations
openssl s_client -connect reallylegacyserver.old -tls1 -cipher DEFAULT:@SECLEVEL=1

works with a server that does not offer pfs. As soon as pfs (ex. ECDHE-RSA-AES256-SHA), it fails and needs seclevel 0. If I understand this correctly this is because the server key exchange signature algorithm is SHA for tls1. But why is in this case AES128-SHA whitelisted without pfs?

### PR DESCRIPTION
**Pull request by paulidale**
This is in line with the NEWS entry (erroneously) announcing such for 3.0.

Fixes #18194

- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by paulidale** [comment]
The failed test is the GOST external test.  Unsurprisingly, that broke with these changes.

**Comment by beldmit** [comment]
It shouldn't be difficult to fix this test enforcing SECLEVEL=0

**Comment by paulidale** [comment]
@beldmit, that's exactly what is required.  I'm not suggesting that the GOST failures should block this.

**Comment by beldmit** [comment]
Will try to do today.

File: ssl/ssl_cert.c [hunk 1]
```diff
@@ -1047,16 +1047,13 @@ static int ssl_security_default_callback(const SSL *s, const SSL_CTX *ctx,
...
+            /* SSLv3, TLS v1.0 and TLS v1.1 only allowed at level 0 */
+            if (nid <= TLS1_1_VERSION && level > 0)
                 return 0;
         } else {
+            /* DTLS v1.0 only allowed at level 0 */
+            if (DTLS_VERSION_LE(nid, DTLS1_VERSION) && level > 0)
+                return 0;
             /* DTLS v1.2 only for level 4 and above */
             if (DTLS_VERSION_LT(nid, DTLS1_2_VERSION) && level >= 4)
                 return 0;
```
Comment (t8m) [inline ssl/ssl_cert.c:1059]: There is no DTLSv1.1 so this `if` does not make any sense anymore.

**Review by t8m — CHANGES_REQUESTED** [review]

File: ssl/ssl_cert.c (hunk 1 above)
Comment (paulidale) [reply ssl/ssl_cert.c]: Oops....

Comment (paulidale) [reply ssl/ssl_cert.c]: Fixed

File: test/recipes/80-test_ssl_old.t [hunk 2]
```diff
@@ -577,44 +615,6 @@ sub testssl {
...
-            ok(!run(test(['ssl_old_test', '-tls1_1', '-no_dhe', '-psk', '0102030405', '-cipher', '@SECLEVEL=3:RSA-PSK-AES256-CBC-SHA384'])),
-               'test auto RSA PSK does not meet security level 3 requirements (PFS)');
-          }
-
-        SKIP: {
-            skip "skipping no PSK at SECLEVEL 3 test", 1
-                if ($no_psk);
-
-            ok(!run(test(['ssl_old_test', '-tls1_1', '-no_dhe', '-psk', '0102030405', '-cipher', '@SECLEVEL=3:PSK-AES256-CBC-SHA384'])),
-               'test auto PSK does not meet security level 3 requirements (PFS)');
```
Comment (mattcaswell) [inline test/recipes/80-test_ssl_old.t:614]: Shouldn't these tests be converted to use TLSv1.2 instead? They don't seem TLSv1.1 specific.

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: test/recipes/80-test_ssl_old.t (hunk 2 above)
Comment (paulidale) [reply test/recipes/80-test_ssl_old.t:614]: Okay.  Done.

**Comment by akarl10** [comment]
correct me if I am wrong, but shouldn't doc/man3/SSL_CTX_set_security_level.pod also be updated?

**Comment by beldmit** [comment]
@paulidale could you please test https://github.com/gost-engine/engine/commit/ee1986c58ccb81d1224d09a7cb56b2043fa6a2e8

**Comment by t8m** [comment]
> correct me if I am wrong, but shouldn't doc/man3/SSL_CTX_set_security_level.pod also be updated?

Good point!

**Comment by beldmit** [comment]
@paulidale could you please try updating the gost-engine submodule to the current master HEAD to fix the tests failure

File: doc/man3/SSL_CTX_set_security_level.pod [hunk 3]
```diff
@@ -80,6 +80,9 @@ cipher suites using CCM with a 64 bit authentication tag are prohibited. Note
 that signatures using SHA1 and MD5 are also forbidden at this level as they
 have less than 80 security bits.
 
+Additionally, SSLv3, TLS 1.0, TLS 1.1 and DTLS 1.0 are all disabled at this
+level.
+
 =item B<Level 2>
 
 Security level set to 112 bits of security. As a result RSA, DSA and DH keys
```
Comment (t8m) [inline doc/man3/SSL_CTX_set_security_level.pod:88]: Please note that there are more notes about disabled protocols at higher levels. These need to be removed.

**Review by t8m — CHANGES_REQUESTED** [review]

File: doc/man3/SSL_CTX_set_security_level.pod (hunk 3 above)
Comment (paulidale) [reply doc/man3/SSL_CTX_set_security_level.pod]: Fixed.

**Comment by paulidale** [comment]
GOST updated and seems to be passing.  Thanks @beldmit for the fix.

**Review by t8m — APPROVED** [review]
Good!

**Review by beldmit — APPROVED** [review]
LGTM

**Comment by paulidale** [comment]
Merged, thanks for the feedback.

**Pull request closed without merging**